### PR TITLE
[#200] Initialize new project directory and copy files recursively

### PR DESCRIPTION
**Pull request by kaungkhantsoe**
[#200](https://github.com/nimblehq/android-templates/issues/200)

## What happened 👀

- Show "=> 🐢 Initializing new project..." in the logs
- Initialize project folder by copying CoroutineTemplate recursively
- Rename project folder according to trimmed app-name
   - If app-name="My Project", then the project folder must be named "MyProject"

## Insight 📝

- Create new folder for the project and name the project without space
- Copy files to new folder recursively
- Enable `gradlew` to be `executable`

## Proof Of Work 📹

https://user-images.githubusercontent.com/32578035/167353504-8e7f22e0-24b2-4a18-a9f3-8f9c5e9dd10e.mov

### CODE REVIEW

File: scripts/new_project.kts [hunk 1]
```diff
@@ -1,25 +1,68 @@
...
+                    showMessage(exception?.message)
+                    return@copyRecursively OnErrorAction.TERMINATE
+                }
+            }
+        }
+    }
 }
 
-NewProject.generate(args)
+NewProject.generate(args)
```
Comment (Tuubz) [inline scripts/new_project.kts:68]: Please keep the new line at the end of the file 🚀

File: scripts/new_project.kts [hunk 2]
```diff
@@ -1,25 +1,68 @@
...
         }
     }
+
+    private fun showMessage(message: String?) {
+        runBlocking {
+            println("\n${message?.trimIndent()}\n")
+        }
+    }
+
+    private suspend fun copyFiles(to: String) {
```
Comment (Tuubz) [inline scripts/new_project.kts:54]: Should we rename `to: String` to `targetPath: String`? 🙏

File: scripts/new_project.kts [hunk 3]
```diff
@@ -1,25 +1,68 @@
+#!/usr/bin/env kscript
+
+@file:MavenRepository("central", "https://repo.maven.apache.org/maven2/")
+@file:DependsOn("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.6.1")
+
+import kotlinx.coroutines.*
+import java.io.File
+import kotlin.coroutines.coroutineContext
```
Comment (Tuubz) [inline scripts/new_project.kts:8]: Could we sort the imports alphabetically? It will look a bit more organized that way 😁

File: scripts/new_project.kts [hunk 4]
```diff
@@ -1,25 +1,68 @@
...
+
 object NewProject {
 
     private const val KEY_APP_NAME = "app-name"
     private const val KEY_PACKAGE_NAME = "package-name"
     private const val ARGUMENT_DELIMITER = "="
+    private const val COROUTINE_FOLDER_NAME = "CoroutineTemplate"
```
Comment (Tuubz) [inline scripts/new_project.kts:15]: Should we keep the naming more general? Ex: `TEMPLATE_FOLDER_NAME`

As most likely we'll rename `CoroutineTemplate` to `template` in the future

File: scripts/new_project.kts [hunk 5]
```diff
@@ -1,25 +1,68 @@
...
+    private const val COROUTINE_PACKAGE_THIRD_SEG = "coroutine"
```
Comment (Tuubz) [inline scripts/new_project.kts:16]: Seems like this is unused, can we remove it? 🐟

File: scripts/new_project.kts [hunk 6]
```diff
@@ -1,25 +1,68 @@
...
         handleArguments(args)
+
+        showMessage("=> 🐢 Initializing new project...")
+        copyFiles(to = appNameWithoutSpace)
+        /*
+         * Set gradlew file as executable
+         * Copying file from one folder to another lost it's previous execution right.
+         * Need to set gradlew as executable to run
+         */
+        File("$appNameWithoutSpace/gradlew")?.setExecutable(true)
```
Comment (Tuubz) [inline scripts/new_project.kts:32]: Shall we put this into a new function, for example: `initializeProjectFolder()` / `initializeProjectDirectory()` ? 🙏

File: scripts/new_project.kts [hunk 7]
```diff
@@ -1,25 +1,68 @@
...
 
     private const val KEY_APP_NAME = "app-name"
     private const val KEY_PACKAGE_NAME = "package-name"
     private const val ARGUMENT_DELIMITER = "="
+    private const val COROUTINE_FOLDER_NAME = "CoroutineTemplate"
+    private const val COROUTINE_PACKAGE_THIRD_SEG = "coroutine"
 
     private var appName = ""
     private var packageName = ""
+    private var appNameWithoutSpace = ""
```
Comment (Tuubz) [inline scripts/new_project.kts:20]: Would it be nicer to have `appNameWithoutSpace` below `appName` 🤩 ?

File: scripts/new_project.kts [hunk 8]
```diff
@@ -1,25 +1,68 @@
...
 
-    fun generate(args: Array<String>) {
+    fun generate(args: Array<String>) = runBlocking {
         handleArguments(args)
+
+        showMessage("=> 🐢 Initializing new project...")
```
Comment (Tuubz) [inline scripts/new_project.kts:25]: Optional: Put it as a `const` -> Example: `MESSAGE_INITIALIZE_PROJECT`

File: scripts/new_project.kts [hunk 9]
```diff
@@ -1,25 +1,68 @@
...
+        runBlocking {
+            println("\n${message?.trimIndent()}\n")
+        }
+    }
+
+    private suspend fun copyFiles(to: String) {
+        withContext(coroutineContext) {
+            async(Dispatchers.IO) {
+                val target = File(to)
+                val file = File("../$COROUTINE_FOLDER_NAME")
```
Comment (Tuubz) [inline scripts/new_project.kts:58]: What about: `val targetFile` & `val sourceFile` 🚀

File: scripts/new_project.kts [hunk 10]
```diff
@@ -1,25 +1,68 @@
...
+    private fun showMessage(message: String?) {
+        runBlocking {
+            println("\n${message?.trimIndent()}\n")
+        }
+    }
+
+    private suspend fun copyFiles(to: String) {
+        withContext(coroutineContext) {
+            async(Dispatchers.IO) {
+                val target = File(to)
```
Comment (Tuubz) [inline scripts/new_project.kts:57]: Should this be `File("../$to")`? 

So we can copy it on to the `root` level instead of inside `scripts` 💭

File: scripts/new_project.kts [hunk 11]
```diff
@@ -1,25 +1,68 @@
+#!/usr/bin/env kscript
+
+@file:MavenRepository("central", "https://repo.maven.apache.org/maven2/")
+@file:DependsOn("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.6.1")
```
Comment (Tuubz) [inline scripts/new_project.kts:4]: Could we also sort these alphabetically? (`@file:DependsOn` in front of `@file:MavenRepository`) 🙇

File: scripts/new_project.kts [hunk 12]
```diff
@@ -1,25 +1,68 @@
...
+
+    private fun showMessage(message: String?) {
+        runBlocking {
+            println("\n${message?.trimIndent()}\n")
+        }
+    }
+
+    private suspend fun copyFiles(to: String) {
+        withContext(coroutineContext) {
+            async(Dispatchers.IO) {
```
Comment (Tuubz) [inline scripts/new_project.kts:56]: By using coroutines, does it mean the script can execute faster than without? 🙏 

If yes, should we move this to `fun generate(...)` instead? So all functions are executed on `Dispatchers.IO`

File: scripts/new_project.kts (hunk 2 above)
Comment (Tuubz) [reply scripts/new_project.kts:54]: What about we merge them together? `toPath`? :sexy_girl

I got inspired by `File.java`:
Screen Shot 2565-05-09 at 17 22 09

File: scripts/new_project.kts (hunk 8 above)
Comment (Tuubz) [reply scripts/new_project.kts:25]: Yeah, no biggie. Although, by moving all of this into a function (ex: `initializeProjectFolder()`), it should be clear already that we're initializing the project 🙏

File: scripts/new_project.kts (hunk 12 above)
Comment (Tuubz) [reply scripts/new_project.kts:56]: Hmmmm, please allow me to search a bit more on this. I'm not so familiar with using coroutines in a script context. Do you have any reference/link I can look into by any chance? 🙏

File: scripts/new_project.kts (hunk 3 above)
Comment (Tuubz) [reply scripts/new_project.kts:7]: @kaungkhantsoe Oops! I believe it should be like this if we want to sort it alphabetically, right?

```
import java.io.File
import kotlin.coroutines.coroutineContext
import kotlinx.coroutines.*
``

File: scripts/new_project.kts [hunk 13]
```diff
@@ -1,25 +1,72 @@
...
+import java.io.File
+
 object NewProject {
 
+    private const val ARGUMENT_DELIMITER = "="
+    private const val COROUTINE_FOLDER_NAME = "CoroutineTemplate"
     private const val KEY_APP_NAME = "app-name"
     private const val KEY_PACKAGE_NAME = "package-name"
-    private const val ARGUMENT_DELIMITER = "="
+    private const val TEMPLATE_FOLDER_NAME = "coroutine"
```
Comment (Tuubz) [inline scripts/new_project.kts:16]: Hmmm, I think there was a misunderstanding here 🙇  My suggestion was to rename `COROUTINE_FOLDER_NAME` to `TEMPLATE_FOLDER_NAME` and remove line 16 as it's not being used anywhere 🙏

File: scripts/new_project.kts [hunk 14]
```diff
@@ -1,25 +1,72 @@
...
         }
     }
+
+    private fun showMessage(message: String?) {
+        runBlocking {
+            println("\n${message?.trimIndent()}\n")
+        }
+    }
+
+    private suspend fun copyFiles(toPath: String) {
```
Comment (Tuubz) [inline scripts/new_project.kts:58]: Should we refactor it to `copyFiles(fromPath: String, toPath: String)`? It will be more flexible this way

`copyFiles(fromPath = "../$COROUTINE_FOLDER_NAME", toPath = "../$appNameWithoutSpace")`

**Review by Tuubz — COMMENTED** [review]
A few more styling/renaming comments from my side 😁

File: scripts/new_project.kts (hunk 12 above)
Comment (manh-t) [reply scripts/new_project.kts:59]: @kaungkhantsoe when we run long-running tasks in an interface, I agree that we should use `IO` thread. But in this case, `copyRecursively` function is using `for loop` for the execution, and other complex tasks are neither executed nor are running on an interface. I think we can get rid of `coroutines` in the script 🙇  Please let me know your opinion too @nimblehq/android-chapter

File: scripts/new_project.kts (hunk 1 above)
Comment (Tuubz) [reply scripts/new_project.kts]: @kaungkhantsoe You can resolve this, by enabling "Ensure every saved file ends with a line break:

Screen Shot 2565-05-11 at 10 21 55

File: scripts/new_project.kts (hunk 12 above)
Comment (Tuubz) [reply scripts/new_project.kts:59]: @kaungkhantsoe Could you please share a screen recording that demonstrates why you chose to go with coroutines? I recall you mentioned to me that you faced an issue if we don't use coroutines 🙏

File: scripts/new_project.kts [hunk 15]
```diff
@@ -1,25 +1,71 @@
...
+                    appName = value
+                    appNameWithoutSpace = appName.trim().replace(" ", "")
+                }
                 KEY_PACKAGE_NAME -> packageName = value
             }
         }
     }
+
+    private fun showMessage(message: String?) {
+        runBlocking {
```
Comment (Wadeewee) [inline scripts/new_project.kts:52]: Do we need to use `runBlocking` here as we already execute show the message under the thread that we created in method `copyFiles` 🤔

**Comment by Wadeewee** [comment]
https://github.com/nimblehq/android-templates/pull/216#discussion_r869831728

@kaungkhantsoe I also agree with @manh-t I don't think we need a suspendable computation for the execution script it should be simple🙏 Please correct me If I missed something 🙇‍♀️

File: scripts/new_project.kts [hunk 16]
```diff
@@ -1,25 +1,55 @@
...
 
+    private const val ARGUMENT_DELIMITER = "="
     private const val KEY_APP_NAME = "app-name"
     private const val KEY_PACKAGE_NAME = "package-name"
-    private const val ARGUMENT_DELIMITER = "="
+    private const val TEMPLATE_FOLDER_NAME = "CoroutineTemplate"
 
     private var appName = ""
+    private val appNameWithoutSpace: String
+        get() = appName.trim().replace(" ", "")
```
Comment (Tuubz) [inline scripts/new_project.kts:14]: Oh, could be it we don't actually need `trim()`, as we are already removing all the whitespaces with  `replace(" ", "")` ? 🙏

File: scripts/new_project.kts [hunk 17]
```diff
@@ -1,25 +1,55 @@
...
+                KEY_APP_NAME -> {
+                    appName = value
+                }
                 KEY_PACKAGE_NAME -> packageName = value
             }
         }
     }
+
+    private fun showMessage(message: String) {
+        println("\n${message.trimIndent()}\n")
```
Comment (Tuubz) [inline scripts/new_project.kts:42]: Do we need `trimIndent()`? 🙏

File: scripts/new_project.kts [hunk 18]
```diff
@@ -1,25 +1,55 @@
...
     }
 
     private fun handleArguments(args: Array<String>) {
         args.forEach {
             val (key, value) = it.split(ARGUMENT_DELIMITER)
             when (key) {
-                KEY_APP_NAME -> appName = value
+                KEY_APP_NAME -> {
+                    appName = value
+                }
```
Comment (Tuubz) [inline scripts/new_project.kts:35]: This could be reverted 🙏

File: scripts/new_project.kts [hunk 19]
```diff
@@ -1,25 +1,55 @@
+#!/usr/bin/env kscript
```
Comment (Tuubz) [inline scripts/new_project.kts:1]: Do we still need this? If I'm not mistaken we no longer need this, unless we import dependencies

File: scripts/new_project.kts [hunk 20]
```diff
@@ -1,25 +1,71 @@
...
     private var packageName = ""
 
-    fun generate(args: Array<String>) {
+    fun generate(args: Array<String>) = runBlocking {
         handleArguments(args)
+
+        initializeNewProjectDirectory()
+    }
+
+    private suspend fun initializeNewProjectDirectory() {
```
[the remark this answers is not in the document]
Comment (Tuubz) [reply scripts/new_project.kts]: @kaungkhantsoe Missed one here 👋 😁

**Review by Tuubz — COMMENTED** [review]
Last one from my side, the rest looks perfect for me ⭐

**Review by Tuubz — APPROVED** [review]
🚀

File: scripts/new_project.kts [hunk 21]
```diff
@@ -1,25 +1,51 @@
...
+        // Set gradlew file as executable, because copying files from one folder to another doesn't copy file permissions correctly (= read, write & execute).
+        File("../$appNameWithoutSpace/gradlew")?.setExecutable(true)
     }
 
     private fun handleArguments(args: Array<String>) {
         args.forEach {
             val (key, value) = it.split(ARGUMENT_DELIMITER)
             when (key) {
-                KEY_APP_NAME -> appName = value
+                KEY_APP_NAME ->  appName = value
```
Comment (Wadeewee) [inline scripts/new_project.kts:31]: Please revert this change format, REST LGTM 🚀

**Review by Wadeewee — APPROVED** [review]

**Comment by Tuubz** [comment]
One last thing here, do you think we should clean up the git commit history a bit? 🙇 

I'm fine if we squash them all into one commit, or be more descriptive

Screen Shot 2565-05-12 at 14 39 53

**Review by hoangnguyen92dn — APPROVED** [review]

**Comment by Tuubz** [comment]
@manh-t One more approval from you and we'll merge this 🚀

**Review by manh-t — APPROVED** [review]
